xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -1186,6 +1178,44 @@ absl::StatusOr<bool> WindowedEinsumHandler::Run(
     changed |= visitor.changed();
   }
 
+  if (all_windowed_einsum_loops.size() > 0) {
+    TF_ASSIGN_OR_RETURN(bool applied_algsimp,
+                        AlgebraicSimplifier(AlgebraicSimplifierOptions())
+                            .Run(module, execution_threads));
+    changed |= applied_algsimp;
+    TF_ASSIGN_OR_RETURN(bool applied_cf,
+                        HloConstantFolding().Run(module, execution_threads));
+    changed |= applied_cf;
+  }
+  for (HloInstruction* loop : all_windowed_einsum_loops) {
+    VLOG(5) << "Processing " << loop->ToString() << " for unrolling.";
+    std::string original_body_name = std::string(loop->while_body()->name());
+    std::string original_cond_name =
+        std::string(loop->while_condition()->name());
+
+    TF_ASSIGN_OR_RETURN(
+        UnrollResult result,
+        WhileLoopUnroller::UnrollAndReturnReplacement(
+            loop, /*unroll_factor=*/-1, /*wrap_in_trivial_loop=*/true,
+            /*force_unroll=*/false));
```

Comment:
Comment the motivation for unrolling the while-loop, and for not allowing it to be inlined.

Also is the motivation to overlap DUS at the end of each iteration with the matmul and/or collective at the start of the next iteration?

Reply:
Yes, this is to allow DUS to overlap with matmul and collective from other iterations, in practice, we actually can overlap the matmul AND DUS from current iteration with matmul and collective from other iterations.

Reply:
Is the DUS codegen-ed as a specific kernel? I would have expected the dot kernel to directly write into the correct place in the buffer, instead of writing to a temp buffer before being copied to the output buffer via DUS.

Reply:
the DUS is code-gened as a part of a fusion kernel fused with some other trivial compute kernels. We have a dot+dus fusion to fuse this but it's not working reliably for collective matmul due to the way the DUS index is computed in the loop, that's being worked on separately from our side. Even with the DUS fused, we still need to unroll to overlap gemms across iterations to minimize wave quantization overhead.

Reply:
Thanks, that all makes sense